simple/utils/spectral_types.py

```diff
@@ -3,320 +3,317 @@
-
-            adopted_ind = source_spt_data["adopted"] == 1
-            if sum(adopted_ind):
-                old_adopted = source_spt_data[adopted_ind]
-                if spectral_type_error[i] < min(source_spt_data["spectral_type_error"]):
+    elif len(source_spt_data) > 0:
+        # Spectral Type Data already exists
+        adopted_ind = source_spt_data["adopted"] == 1
+        if sum(adopted_ind):
+            old_adopted = source_spt_data[adopted_ind]
+            logger.debug(f"Old adopted data: {old_adopted}")
+            if (
+                old_adopted["spectral_type_error"] is not None
+                and spectral_type_error is not None
+            ):
+                if spectral_type_error < min(old_adopted["spectral_type_error"]):
                     adopted = True
-
-                    if old_adopted:
-                        with db.engine.connect() as conn:
-                            conn.execute(
-                                db.SpectralTypes.update()
-                                .where(
-                                    and_(
-                                        db.SpectralTypes.c.source
-                                        == old_adopted["source"][0],
-                                        db.SpectralTypes.c.reference
-                                        == old_adopted["reference"][0],
-                                    )
-                                )
-                                .values(adopted=False)
-                            )
-                            conn.commit()
-                        # check that adopted flag is successfully changed
-                        old_adopted_data = (
-                            db.query(db.SpectralTypes)
-                            .filter(
-                                and_(
-                                    db.SpectralTypes.c.source
-                                    == old_adopted["source"][0],
-                                    db.SpectralTypes.c.reference
-                                    == old_adopted["reference"][0],
-                                )
-                            )
-                            .table()
-                        )
-                        logger.debug("Old adopted measurement unset")
-                        if logger.level == 10:
-                            old_adopted_data.pprint_all()
-
+                else:
+                    adopted = False
                 logger.debug(f"The new spectral type's adopted flag is:, {adopted}")
+            else:
+                adopted = True
+                logger.debug(
+                    "No spectral type error found, setting adopted flag to True"
+                )
         else:
-            msg = "Unexpected state"
-            logger.error(msg)
-            raise RuntimeError
+            adopted = True
+            old_adopted = None
+            logger.debug("No adopted data found, setting adopted flag to True")
```

Comment:
I'm OK with this for now BUT this screams for a helper function that solely handles the determination of whether adopted should be set or not. As a helper function, it can be independently tested and implemented anywhere we have adopted columns.

Reply:
And to clarify- not updating the database, just determining and returning the values of adopted/old_adopted.

Reply:
I at least made a helper function just for spectral type. Every adopted column is gonna have different logic so making a generic one will be hard.

Reply:
Yep, this is good for now. I was thinking that it would take as input `source_spt_data` so it wouldn't need to know how to query the database (or do any DB access at all), but since there are some other if-statements relevant to spectral types it's worth tackling this on its own.
I created an issue for a more generalized set of functions: https://github.com/SIMPLE-AstroDB/SIMPLE-db/issues/542